src/gui/vectortile/qgsvectortilebasiclabelingwidget.cpp

```diff
@@ -0,0 +1,458 @@
+  : QgsMapLayerConfigWidget( layer, canvas, parent )
+  , mVTLayer( layer )
+  , mMessageBar( messageBar )
+{
+
+  setupUi( this );
+  layout()->setContentsMargins( 0, 0, 0, 0 );
+
+  if ( layer->labeling() && layer->labeling()->type() == QStringLiteral( "basic" ) )
+  {
+    mLabeling.reset( static_cast<QgsVectorTileBasicLabeling *>( layer->labeling()->clone() ) );
+  }
+  else
+  {
+    mLabeling.reset( new QgsVectorTileBasicLabeling() );
+  }
+
+  mModel = new QgsVectorTileBasicLabelingListModel( mLabeling.get(), viewStyles );
+  viewStyles->setModel( mModel );
+
+  QMenu *menuAddRule = new QMenu( btnAddRule );
+  menuAddRule->addAction( tr( "Marker" ), this, [this] { addStyle( QgsWkbTypes::PointGeometry ); } );
+  menuAddRule->addAction( tr( "Line" ), this, [this] { addStyle( QgsWkbTypes::LineGeometry ); } );
+  menuAddRule->addAction( tr( "Fill" ), this, [this] { addStyle( QgsWkbTypes::PolygonGeometry ); } );
+  btnAddRule->setMenu( menuAddRule );
+
+  //connect( btnAddRule, &QPushButton::clicked, this, &QgsVectorTileBasicLabelingWidget::addStyle );
+  connect( btnEditRule, &QPushButton::clicked, this, &QgsVectorTileBasicLabelingWidget::editStyle );
+  connect( btnRemoveRule, &QAbstractButton::clicked, this, &QgsVectorTileBasicLabelingWidget::removeStyle );
+
+  connect( viewStyles, &QAbstractItemView::doubleClicked, this, &QgsVectorTileBasicLabelingWidget::editStyleAtIndex );
+
+  connect( mModel, &QAbstractItemModel::dataChanged, this, &QgsPanelWidget::widgetChanged );
+  connect( mModel, &QAbstractItemModel::rowsInserted, this, &QgsPanelWidget::widgetChanged );
+  connect( mModel, &QAbstractItemModel::rowsRemoved, this, &QgsPanelWidget::widgetChanged );
+}
+
+QgsVectorTileBasicLabelingWidget::~QgsVectorTileBasicLabelingWidget() = default;
+
+void QgsVectorTileBasicLabelingWidget::apply()
+{
+  mVTLayer->setLabeling( mLabeling->clone() );
+}
+
+void QgsVectorTileBasicLabelingWidget::addStyle( QgsWkbTypes::GeometryType geomType )
+{
+  QgsVectorTileBasicLabelingStyle style;
+  style.setGeometryType( geomType );
+
+  int rows = mModel->rowCount();
+  mModel->insertStyle( rows, style );
+  viewStyles->selectionModel()->setCurrentIndex( mModel->index( rows, 0 ), QItemSelectionModel::ClearAndSelect );
+}
+
+void QgsVectorTileBasicLabelingWidget::editStyle()
+{
+  editStyleAtIndex( viewStyles->selectionModel()->currentIndex() );
+}
+
+void QgsVectorTileBasicLabelingWidget::editStyleAtIndex( const QModelIndex &index )
+{
+  QgsVectorTileBasicLabelingStyle style = mLabeling->style( index.row() );
+
+  QgsPalLayerSettings labelSettings = style.labelSettings();
+  if ( labelSettings.layerType == QgsWkbTypes::UnknownGeometry )
+    labelSettings.layerType = style.geometryType();
+
+  QgsSymbolWidgetContext context;
+  context.setMapCanvas( mMapCanvas );
+  context.setMessageBar( mMessageBar );
+
+  QgsVectorLayer *vectorLayer = nullptr;  // TODO: have a temporary vector layer with sub-layer's fields?
```

Comment:
I think you'll need this in order to get the expression context correctly setup for the labeling widget

Reply:
Yeah, I have a similar to-do item for the "basic renderer" widget, and a bunch of other bits to make the GUI more user friendly. If you don't mind I would address those things afterwards during feature freeze.